### PR TITLE
[CUDA] Tune hardware-aware split-K decode scheduling

### PR DESCRIPTION
**Pull request by Gasoonjia**
## Stack dependency

Depends on #22189.

## Summary

Combine the D113 SM-sized split-K work with the final scheduling policy for the numerically stable online-softmax implementation.

- Derive split counts from the target GPU SM count, `L_kv`, and `B * H_kv`.
- Keep the launch size static while reading runtime `kv_len` on device, preserving CUDA graph compatibility.
- Retune the occupancy target from D113's two waves to 16/9 waves.
- Cap splits by both KV-buffer size and a maximum of 128 partitions to avoid excessive reduction overhead on high-SM GPUs.
- Use contiguous KV partitions instead of D113's persistent strided traversal; profiling showed that contiguous partitions are faster for the online-softmax kernel.
- Apply the same hardware-aware policy to both decode and small-query split-K paths.
- Keep scale arithmetic explicitly in FP32.

The policy contains no model-specific or GPU-specific shape hard-coding. The SM count is specialized during native export. Under the supported deployment flow, each target architecture is exported independently and those variants may then be merged into one PTE; moving a native artifact to a same-architecture GPU with a different SM count remains correct, although its scheduling can be suboptimal.

## Operator-level performance

RTX 5090 CUDA Graph replay, BF16 GQA decode (`B=1, Hq=32, Hkv=2, D=128`), compared with #22189:

| KV buffer | Runtime KV | #22189 | This PR | Change |
|---:|---:|---:|---:|---:|
| 4,096 | 512 | 10.86 us | 9.99 us | +8.6% |
| 4,096 | 2,048 | 12.66 us | 10.66 us | +18.8% |
| 4,096 | 4,096 | 13.49 us | 12.51 us | +7.8% |
| 130,048 | 32,768 | 33.21 us | 33.13 us | +0.2% |
| 130,048 | 130,048 | 95.50 us | 96.50 us | -1.0% |

The optimization primarily improves the sliding-window decode family. The 128-split cap keeps the long global-attention path within approximately 1% of the preceding PR instead of over-splitting it.

## Test plan

- `lintrunner init && lintrunner -a` (clean).
- Triton split-K and FP8 dispatch suite: 47 passed.
- Standalone D113 decode correctness smoke: 3 passed.
- Covered runtime KV-length tail exclusion, non-power-of-two split counts, and large-logit stability.
- `git diff --check`.

### CODE REVIEW

File: backends/cuda/triton/kernels/sdpa.py [hunk 1]
```diff
@@ -51,8 +52,70 @@ def _is_power_of_2(n: int) -> bool:
...
+# not depend on floating-point rounding.
+_SPLITK_TARGET_WAVES_NUMERATOR = 16
+_SPLITK_TARGET_WAVES_DENOMINATOR = 9
+_SPLITK_MAX_SPLITS = 128
+
+# Do not create more static split CTAs than one per 128 elements in the KV
+# buffer. This permits 16 splits for the 2048-element sliding-window family.
+# A 40-state, >L2 working-set A/B showed gains at the production prompt lengths;
+# single-state measurements are cache-hot and are not representative here.
+# The occupancy cap independently lowers the 128K global family from 85 to 76
```
Comment (shoumikhin) [inline backends/cuda/triton/kernels/sdpa.py]: Where does 85 come from? On the merge base the split count is `min(max(cdiv(L_kv, 256), 1), 128)`, a pure buffer-length rule that returns 128 for the 128K case, not 85. The arithmetic of "85 to 76 on a 170-SM device" does reproduce with the new formula, so I think this describes an intermediate version rather than previous behaviour. Worth rewording so the next reader is not hunting for a baseline that never shipped.

File: backends/cuda/triton/kernels/sdpa.py [hunk 2]
```diff
@@ -51,8 +52,70 @@ def _is_power_of_2(n: int) -> bool:
...
+# splits on a 170-SM device.
+_SPLITK_BUFFER_ELEMENTS_PER_SPLIT = 128
+
+
+@functools.lru_cache(maxsize=None)
+def _device_sm_count(device_index: int) -> int:
+    """SM (multiprocessor) count of a CUDA device.
+
+    Read once per device from torch device properties. This is a fixed
+    hardware property, so it is safe to bake into the split count at
```
Comment (shoumikhin) [inline backends/cuda/triton/kernels/sdpa.py]: "Safe to bake into the split count at export" holds within one machine, not across export host and run target. The SM count ends up as a literal in the generated wrapper, with the partial buffers sized to match. Packaging in this repo ships artifacts per compute capability, not per SM count, so an SM90 artifact can land on an SM90 device with a different SM count and run a split count tuned for the wrong machine.

Performance only, and I do not think it blocks the PR. But the wording tells the next reader a cross-machine assumption is safe, while `cuda_backend.py` already treats `TORCH_CUDA_ARCH_LIST` as the way to name a different target.

**Review by shoumikhin — COMMENTED** [review]
No correctness problem found here. The 16/9 wave target, the buffer cap and the 128 clamp all behave as described, and the split count really is fixed before graph capture.

Two comment-level notes inline: one cites a previous baseline that does not match the merge base, and one calls a cross-machine assumption safe. Neither blocks landing.

File: backends/cuda/triton/kernels/sdpa.py [hunk 3]
```diff
@@ -51,8 +52,71 @@ def _is_power_of_2(n: int) -> bool:
...
-def _decode_splitk_config(L_kv: int) -> tuple[int, int]:
-    num_splits = min(max(triton.cdiv(L_kv, _SPLITK_LKV_THRESHOLD), 1), 128)
+# Decode split-K occupancy target. A sweep across both production attention
+# families showed that targeting 16/9 waves gives a good balance between split
+# kernel occupancy and reduction/empty-CTA overhead. Keep the ratio integral so
+# the launch policy is deterministic and does not depend on floating-point
+# rounding.
+_SPLITK_TARGET_WAVES_NUMERATOR = 16
+_SPLITK_TARGET_WAVES_DENOMINATOR = 9
+_SPLITK_MAX_SPLITS = 128
```
Comment (digantdesai) [inline backends/cuda/triton/kernels/sdpa.py:62]: just more magic? like max_splits for a large Lkv on a large GPU could be too small?

File: backends/cuda/triton/kernels/sdpa.py [hunk 4]
```diff
@@ -51,8 +52,71 @@ def _is_power_of_2(n: int) -> bool:
...
+            _SPLITK_TARGET_WAVES_DENOMINATOR * max(grid_y, 1),
+        ),
+        1,
+    )
+    buffer_cap = max(triton.cdiv(L_kv, _SPLITK_BUFFER_ELEMENTS_PER_SPLIT), 1)
+    # Avoid letting very high-SM GPUs over-split long global-attention buffers:
+    # beyond 128 partitions the extra reduction and launch work outweighed the
+    # occupancy gain in CUDA-graph measurements. Lower-SM devices continue to
+    # use their hardware-derived cap.
+    num_splits = min(buffer_cap, sm_cap, _SPLITK_MAX_SPLITS)
```
Comment (digantdesai) [inline backends/cuda/triton/kernels/sdpa.py:119]: how does llama.cpp do this? is there a way to make this a bit less sensitive to sm count?

**Review by digantdesai — APPROVED** [review]

File: backends/cuda/triton/kernels/sdpa.py (hunk 1 above)
Comment (Gasoonjia) [reply backends/cuda/triton/kernels/sdpa.py]: The 85 value came from an intermediate implementation. I reworded the comment to compare against the merge-base cap of 128: the new policy selects 76 splits for the stated 170-SM/grid-y=4 case.

File: backends/cuda/triton/kernels/sdpa.py (hunk 2 above)
Comment (Gasoonjia) [reply backends/cuda/triton/kernels/sdpa.py]: Good point. I changed the wording to state that the SM count is specialized into each native artifact during export. Running that artifact on a same-architecture GPU with a different SM count remains correct but may be suboptimal. Our supported cross-architecture flow exports native variants independently and merges them via #22198, or uses the PTX fallback path.
